megamek/src/megamek/utilities/ai/DatasetParser.java

```diff
@@ -0,0 +1,196 @@
+ * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
+ * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+ * for more details.
+ *
+ */
+package megamek.utilities.ai;
+
+import megamek.common.*;
+import megamek.common.util.StringUtil;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Parses a dataset from one or more files and turns it into a training dataset.
+ * The dataset currently expected is the game_actions_log.tsv file generated by the megamek client hosting the game if the
+ * option is enabled
+ * @author Luana Coppio
+ */
+public class DatasetParser {
+
+    // attackAction is not being considered right now
+    private final List<ActionAndState> actionAndStates = new ArrayList<>();
+    private final Map<Integer, Entity> entities = new HashMap<>();
+    private int idOffset = 0;
+    private int highestEntityId = 0;
+
+    /**
+     * Parses a dataset from a file. Can be chained with other parse calls to create a large single training dataset.
+     * @param file The file to parse
+     * @return The parser instance
+     */
+    public DatasetParser parse(File file) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (line.startsWith("PLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tFACING\tFROM_X\tFROM_Y\tTO_X\tTO_Y\tHEXES_MOVED\tDISTANCE\tMP_USED\tMAX_MP\tMP_P\tHEAT_P\tARMOR_P\tINTERNAL_P\tJUMPING\tPRONE\tLEGAL\tSTEPS")) {
+                    // Parse action line
+                    String actionLine = reader.readLine();
+                    if (actionLine == null) break;
+                    UnitAction action = parseActionLine(actionLine);
+
+                    // Parse state block
+                    line = reader.readLine(); // State header
+                    if (line == null || !line.startsWith("ROUND\tPHASE\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
+                        // || !line.startsWith("ROUND\tPHASE\tTEAM_ID\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
+
+                        if (line == null // || !line.startsWith("ROUND\tPHASE\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
+                            || !line.startsWith("ROUND\tPHASE\tTEAM_ID\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
```

Comment:
Same here.

Reply:
I see no benefit into making this an enum, care to elaborate?

Reply:
Yeah, apologies, I was thinking this was a set of matching strings, not one huge string.
It would still benefit from being made into a constant for clarity.